docs/cudf/source/developer_guide/library_design.md

```diff
@@ -325,30 +325,26 @@ This section describes the internal implementation details of the copy-on-write
 It is recommended that developers familiarize themselves with [the user-facing documentation](copy-on-write-user-doc) of this functionality before reading through the internals
 below.
 
-The core copy-on-write implementation relies on the `CopyOnWriteBuffer` class.
-When the cudf option `"copy_on_write"` is `True`, `as_buffer` will always return a `CopyOnWriteBuffer`.
-This subclass of `cudf.Buffer` contains all the mechanisms to enable copy-on-write behavior.
-The class stores [weak references](https://docs.python.org/3/library/weakref.html) to every existing `CopyOnWriteBuffer` in `CopyOnWriteBuffer._instances`, a mapping from `ptr` keys to `WeakSet`s containing references to `CopyOnWriteBuffer` objects.
-This means that all `CopyOnWriteBuffer`s that point to the same device memory are contained in the same `WeakSet` (corresponding to the same `ptr` key) in `CopyOnWriteBuffer._instances`.
-This data structure is then used to determine whether or not to make a copy when a write operation is performed on a `Column` (see below).
-If multiple buffers point to the same underlying memory, then a copy must be made whenever a modification is attempted.
+The core copy-on-write implementation relies on the factory function `as_tenable_buffer` and the two classes `TenableBuffer` and `BufferSlice`.
+
+`TenableBuffer` is a subclass of the regular `Buffer` that tracks its "expose" status of its underlying memory. We say that the buffer has been exposed if the device pointer (integer or void*) has been accessed outside of cudf, in which case we have no control over knowing if the data is being modified by a third-party. Additionally, `TenableBuffer` also maintains [weak references](https://docs.python.org/3/library/weakref.html) to every existing `BufferSlice` that points to its underlying memory.
+
+`BufferSlice` is a subclass of `TenableBuffer` that represents a _slice_ of the memory underlying a tenable buffer.
```

Comment:
Is subclassing the right thing? (I haven't yet read the implementation). It seems like every `BufferSlice` `has-a` `TenableBuffer`, but that doesn't necessarily imply an `is-a` relationship.

If it really is a subclass, do we actually need both classes, or can there just be `TenableBuffer` objects that either own data, or are views of existing data.

Reply:
In a follow-up PR, `SpillableBuffer` will inherent from `TenableBuffer` and `BufferSlice._base` will point to `SpillableBuffer`.

Reply:
Maybe it is better if `BufferSlice` inherent from `Buffer` ?

Reply:
It kind of seems like what we want is [ad-hoc](https://en.wikipedia.org/wiki/Ad_hoc_polymorphism), rather than [subtype](https://en.wikipedia.org/wiki/Subtyping), polymorphism (`BufferSlice`s are not Liskov-substitutable for `TenableBuffer`s everywhere [in particular when constructing new buffer slices]). In which case, is this the time to introduce a buffer [`Protocol`](https://peps.python.org/pep-0544/)?

Reply:
A buffer Protocol might make sense, but then I think we should do it in a follow up PR

Reply:
Alternatively, we could have `Buffer` use `BufferSlice` so that `BufferSlice` is the only buffer object used in the rest of cudf.
In any case, I think we should wait until after the next PR that makes `SpillableBuffer` use  `BufferSlice` and `TenableBuffer`